flink-cdc-base/src/main/java/com/ververica/cdc/connectors/base/source/reader/IncrementalSourceReader.java

```diff
@@ -129,12 +136,27 @@ public List<SourceSplitBase> snapshotState(long checkpointId) {
         // add stream splits who are uncompleted
         stateSplits.addAll(uncompletedStreamSplits.values());
 
+        for (SourceSplitBase split : stateSplits) {
+            if (split.isStreamSplit()) {
+                lastCheckPointStreamSplit.put(checkpointId, split.asStreamSplit());
+                LOG.debug(
+                        "snapshot stream split, checkpoint id {}, stream split {}",
+                        checkpointId,
+                        split.asStreamSplit());
+            }
+        }
+
         return stateSplits;
     }
 
     @Override
     public void notifyCheckpointComplete(long checkpointId) throws Exception {
-        dialect.notifyCheckpointComplete(checkpointId);
+        if (checkpointId > maxCompletedCheckpointId) {
+            SourceSplitBase split = lastCheckPointStreamSplit.get(checkpointId);
+            dialect.notifyCheckpointComplete(checkpointId, split);
+            lastCheckPointStreamSplit.headMap(checkpointId).clear();
+            maxCompletedCheckpointId = checkpointId;
+        }
     }
 
     @Override
```

Comment:
Remove finished splits from lastCheckPointStreamSplit, nor memory leak will occur.

Reply:
Thanks for your review. Yes, StreamSplit of current checkpointId should remove.